pkg/runtime/core/trainingruntime.go

```diff
@@ -187,24 +187,100 @@ func (r *TrainingRuntime) mergePodSpecOverrides(trainJob *trainer.TrainJob, jobS
 			}) {
 				continue
 			}
-			patch, err := json.Marshal(podSpecOverride)
+
+			// Build a sparse JSON patch that omits untouched fields (notably, omit spec.containers unless overridden)
+			podTemplatePatch := map[string]any{}
+			if len(podSpecOverride.Labels) > 0 || len(podSpecOverride.Annotations) > 0 {
+				md := map[string]any{}
+				if len(podSpecOverride.Labels) > 0 {
+					md["labels"] = podSpecOverride.Labels
+				}
+				if len(podSpecOverride.Annotations) > 0 {
+					md["annotations"] = podSpecOverride.Annotations
+				}
+				podTemplatePatch["metadata"] = md
+			}
+
```

Comment:
@xigang @astefanutti @tenzen-y @Electronic-Waste What if we always set the podTemplatePatch from podSpecOverride without checking if values exist ?

```go
podTemplatePatch["metadata"] = podSpecOverride
podTemplatePatch["spec"] = podSpecOverride
```
Would the patch work ?


Right now, we apply patch directly without any [changes to the PodSpecOverride API](https://github.com/kubeflow/trainer/blob/master/pkg/runtime/core/trainingruntime.go#L190C31-L190C46). Although, it has `targetJobs` within its API.

Reply:
Yes, sorry, I just meant what @astefanutti said.